LFS-1160: Deleting the variants file after using the variants uploader triggers an error

IntegralProgrammer opened the pull request:
This PR introduces the following changes:

- Trim the trailing slash from the URL that opens the new Form when within the variants uploader
- Ensure that all network access done by a Form to said Form's JCR node does not contain a trailing slash

To test:

1. Build this branch (`git checkout LFS-1160`, `mvn clean install`)
2. Start CARDS with the `lfs` runMode enabled
3. Create a fake Variants CSV file (`echo "Hello World" > AB12345_1.csv`)
4. Upload this CSV file through the _Variants_ uploader
5. Mouse over the _Somatic Variants_ link for the uploaded file. Verify that its target URL does not contain a trailing slash
6. Click the link
7. Click the _Edit_ pencil button
8. Click the trash can icon to delete the uploaded CSV file
9. Click _Save_. The Form should save properly

Test part two:

10. Repeat the above test but after getting to _step 6_, add a trailing slash to the loaded webpage (browser address bar) and reload. Ensure that the rest of the test completes properly.

IntegralProgrammer committed:
LFS-1160: Deleting the variants file after using the variants uploader triggers an error

Trim the trailing slash from the URL that opens the new Form when within the variants uploader

Ensure that all network access done by a Form to said Form's JCR node does not contain a trailing slash

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -113,7 +113,7 @@ function Form (props) {
   let pageNameWriter = usePageNameWriterContext();
   const history = useHistory();
-  const formURL = `/Forms/${id}`;
+  const formURL = `/Forms/${id}`.replace(/\/$/, "");
```
Rather than removing the trailing `/` here, we should ensure that the `id` prop is truly the Form's (clean) identifier, so the value passed to the Form without a trailing slash.

marta- replied:
Adjusting this regexp should do it: https://github.com/ccmbioinfo/lfs/blob/dev/modules/data-entry/src/main/frontend/src/dataHomepage/Forms.jsx#L34

IntegralProgrammer replied:
Yes, I adjusted that regexp and that fixed the issue. :heavy_check_mark:

sashaandjic commented:
**tested** - waiting for refactoring to do the **re-test**

IntegralProgrammer committed:
LFS-1160: Deleting the variants file after using the variants uploader triggers an error

Address code review comment - instead of removing trailing slash when creating formURL in Form.jsx, do not capture it when creating entry/id in Forms.jsx

marta- commented on modules/variants/src/main/frontend/src/variantFilesContainer.jsx:
```diff
@@ -841,7 +841,7 @@ export default function VariantFilesContainer() {
                     {tumorSubjectLabel} <Link href={tumorPath} target="_blank"> {file.tumor.id} </Link>
                     { file?.region?.path && <> / {regionSubjectLabel} <Link href={regionPath} target="_blank"> {file.region.id} </Link> </> }
-                    { file.formPath && <> : <Link href={file.formPath.replace("/Forms", "Forms")} target="_blank">{somaticVariantsTitle}</Link> </>}
+                    { file.formPath && <> : <Link href={file.formPath.replace("/Forms", "Forms").replace(/\/$/, "")} target="_blank">{somaticVariantsTitle}</Link> </>}
```
Is this change still needed?

IntegralProgrammer replied:
Without it, clicking the _Somatic Variants_ link goes to `/content.html/Forms/THE-FORM-UUID/` (note the trailing slash). While the trailing slash is removed when the _Edit_ (pencil) button is clicked, and therefore the LFS-1160 bug is fixed, I still like this change as it removes the trailing slash when opening the Form and is thus consistent with what happens when you click the Form link on the main _Dashboard_ page.

marta- replied:
Ok, then let's remove the trailing `/` "at the source", on line [681](https://github.com/ccmbioinfo/lfs/blob/dev/modules/variants/src/main/frontend/src/variantFilesContainer.jsx#L681) - unless I'm missing something, this path doesn't appear to be used anywhere else but in this link.

And while we're cleaning up, this also raises the question why, when assigning `file.formPath` on line 681, we prefix it with a `/`, only to remove it here via `file.formPath.replace("/Forms", "Forms")`.

IntegralProgrammer replied:
What are your thoughts on my latest commit `a90e2b2` as a solution thus far? I still would need to address the issue of prefixing `file.formPath` with a `/` only to remove it later on.

marta- replied:
Looks fine to me.

IntegralProgrammer committed:
LFS-1160: Deleting the variants file after using the variants uploader triggers an error

Address code review comment - do not append a trailing slash to formPath or answerPath but rather add the slash when adding any descendant nodes

sashaandjic commented:
Ready for testing?

marta- commented:
> Ready for testing?

This part hasn't been addressed yet, but one could argue it's code optimization and out of the scope of 1160:

> when assigning file.formPath on line 681, we prefix it with a /, only to remove it here via file.formPath.replace("/Forms", "Forms")

IntegralProgrammer committed:
LFS-1160: Deleting the variants file after using the variants uploader triggers an error

Address code review comment - do not prefix file.formPath with a leading '/' only to remove it with file.formPath.replace("/Forms", "Forms")

sashaandjic reviewed (approved):
RE-TESTED
